### ISSUE TITLE
database handling should not be tightly coupled to mongo

### ISSUE THREAD

**Issue by bitterulf** [pencilblue/pencilblue#553]
Not everyone want to deal with mongo for just a small blog or has the ability to host it.
For demonstration, testing and small sites it would be awesome to switch from mongo to nedb.

At the moment this is not possible without some nasty hacks (nedb is mostly mongo compatible but uses another id format).

If the db access would be more abstract like the media provider user can create third party extension to store the database in whatever place as long as they implement the right interface.

**Comment by blakecallens** [comment]
This is in our roadmap, and we're already working towards it, thought it's a long term project for us. You'll notice that in v0.4.0, we removed the use of MongoDB's _id variable from controllers and replaced it with pb.DAO.getIdField().

This allows the DAO to be the single point of entry into a DB system. The goal is for plugins to be able to switch out the DB system, so PencilBlue can be used with just about any system.

**Comment by miguel12345** [comment]
Hi,

I'm currently using Sequelize as my DAO/ORM and I really need something great for the frontend to manage the content. All the other CMS's out there seem to be completely locked in to Mongo, but this issue gives me some hope to use PencilBlue to managed by postgres-sequelize data. 

What is the current status on this? Is it possible to use a custom DAO/ORM without losing hours/days?

**Comment by brianhyder** [comment]
@miguel12345 We've never tried to integrate sequelize into the platform so I can't say with confidence how hard it would be.  We know that the project is large.  Most of that reason is due to the fact that Mongo uses "_id" rather than a traditional "id" variable for each document.  

Our long term vision is to slowly convert all object interactions into ```BaseObjectService``` implementations to prepare the platform for a change of this magnitude.  The BaseObjectService abstracts the majority of the CRUD operations.  

Once all of that refactor is in place we'll look to start phasing out the "_id" properties and move to a standard "id" so that transitioning to another DB would be easier.  It will take some time to do that and to remove references to ```_id``` from the UI.  

The other pain point is that we are introducing multi-site capabilities in 0.5.0.  When this goes in we would need to figure out how to support the extra "site" column on the majority of the objects so that they are scoped to a particular site or "global" when in single site mode.  So there are a more than a few hurdles that we are trying to overcome along the way.  

Realistically, I don't see the project providing multi-database support in 2015 but would be a potential target for 2016.

**Comment by Karnith** [comment]
Hi,
I know this thread is old. Is there any progress on this? Has thought been put into using a system like bookshelf and knex for multi-database support?

**Comment by brianhyder** [comment]
Great question.  The short answers are yes and no, respectively.  We have started to prep the platform for multiple DB support by taking a layered approach to the code base (controller, service(s), DAO).  The services that interact with system collections extend a common service that allows for all of the data abstraction to the DAO layer.  The DAO prototype provides direct access to DB collections.  Using this abstraction we are able to do things such as replace mongo's "_id" field with a more common "id" field.  Additionally, we can provide hooks into the CRUD operations to manipulate queries and results.  

We have not looked into packages such as bookshelf or knex.  The reason is because PB was built on top of a NoSQL solution.  Supporting both SQL and NoSQL would prove quite difficult.  In fact, it has been hard enough to wade through the murky waters of supporting multiple DBs of similar nature.  The next DB that we would likely target is CouchBase.  We are looking at it for a couple of reasons.  First, it is highly scalable and has good UI tools built in to accomplish this.  It aligns with our mission to make PB enterprise grade and horizontally scalable.   Secondly, CouchBase is built on top of memcache.  This would allow the DB solution to act as the cache and the DB.  

In summary, we will most likely stick to NoSQL solutions for the foreseeable future.  It more easily allows for our architecture to allow developers to build plugins that provide additional fields and structures to existing system objects.
